### DISCUSSION TITLE
Registry return out of memory to download modules

### DISCUSSION THREAD

**Discussion by klinux** [terrakube-io/terrakube#1575] [General]
Hello guys,

After upgrade terrakube we start to receive this error to download modules from registry.

```
terrakube-registry java.lang.OutOfMemoryError: Cannot reserve 6226740 bytes of direct buffer memory (allocated: 6291120, limit: 10485760)
terrakube-registry-555496bc9c-7vzmk terrakube-registry  at java.base/java.nio.Bits.reserveMemory(Unknown Source) ~[na:na]
terrakube-registry-555496bc9c-7vzmk terrakube-registry  at java.base/java.nio.DirectByteBuffer.<init>(Unknown Source) ~[na:na]
terrakube-registry-555496bc9c-7vzmk terrakube-registry  at java.base/java.nio.ByteBuffer.allocateDirect(Unknown Source) ~[na:na]
terrakube-registry-555496bc9c-7vzmk terrakube-registry  at java.base/sun.nio.ch.Util.getTemporaryDirectBuffer(Unknown Source) ~[na:na]
...
```

Latest stable version

**Comment by alfespa17** [comment]
Hello @klinux 

Maybe you can try giving the registry pod more memory.

You can also tweak the JVM parameter related to memory using the following:

https://github.com/AzBuilder/terrakube/issues/1499#issuecomment-2491767855

I hope that can helps

**Comment by klinux** [comment]
I think that registry has  a memory leak or other problems, I put a lot of memory and error still persists.

**Reply by alfespa17** [reply]
Can you provide more information about the error, the logs, the version that you are using, the steps to reproduce the error?  It will also help to understand the storage backend that your are using for the registry

**Reply by alfespa17** [reply]
There is a second image that are based on alpaquita Linux that you can try just for the registry 

https://hub.docker.com/layers/azbuilder/open-registry/2.23.3-alpaquita/images/sha256-9e46e5f57d3a4f8f9e3c4f1fee76037753efbe924b3609e80297851d3ecc3636?context=explore

In theory is memory optimized because it is using a special Linux distribution designed for Java apps. Maybe that helps

**Reply by klinux** [reply]
Sure,

Version: 2.23.3
Storage: GCS

I restarted the registry, I can download the module, but after 5, 6 times of the download, I started to receive this error:

```
terrakube-registry java.lang.OutOfMemoryError: Cannot reserve 6226740 bytes of direct buffer memory (allocated: 6291120, limit: 10485760)
terrakube-registry-555496bc9c-7vzmk terrakube-registry  at java.base/java.nio.Bits.reserveMemory(Unknown Source) ~[na:na]
terrakube-registry-555496bc9c-7vzmk terrakube-registry  at java.base/java.nio.DirectByteBuffer.<init>(Unknown Source) ~[na:na]
terrakube-registry-555496bc9c-7vzmk terrakube-registry  at java.base/java.nio.ByteBuffer.allocateDirect(Unknown Source) ~[na:na]
terrakube-registry-555496bc9c-7vzmk terrakube-registry  at java.base/sun.nio.ch.Util.getTemporaryDirectBuffer(Unknown Source) ~[na:na]
...
```
The module has a lot update, is a new module that we created, looks like after update the module this error happens, it's make sense?

I raise the pod RAM, to 6GB, change de parameter **MaxDirectMemorySize** to 32M now I'm testing with 512M, with 32M the error still happens.

Thank you

**Reply by alfespa17** [reply]
I will try to deploy the registry using a GCS tomorrow maybe I can reproduce your error, we update some of the libraries related to Google Cloud according to the release notes not sure if that could be related. 

https://github.com/AzBuilder/terrakube/releases/tag/2.23.0

Can you connect to the GCS and check the size of the module? (maybe the module is to big)

**Reply by alfespa17** [reply]
When you create a new module tag terrakube clone the repository using the tag and creates a new zip file that is downloaded later from the terraform cli, maybe the zip file is too big.

**Reply by klinux** [reply]
At bucket, the module has 19MB, hummmm I think there some problems here.

**Reply by klinux** [reply]
The problem is the size of the module, someone committed the node_modules folder :(

Thank you
